dev/Installer/src/install.cpp

```diff
@@ -0,0 +1,172 @@
+    THROW_IF_FAILED(id->GetPackageFamilyName(&properties->family_name));
+    APPX_PACKAGE_ARCHITECTURE arch;
+    THROW_IF_FAILED(id->GetArchitecture(&arch));
+    properties->architecture = static_cast<ProcessorArchitecture>(arch);
+    THROW_IF_FAILED(id->GetVersion(&properties->version));
+
+    return properties;
+}
+
+wil::com_ptr<IStream> open_temp_file(_In_ PCWSTR path)
+{
+    wil::com_ptr<IStream> outstream;
+    THROW_IF_FAILED(SHCreateStreamOnFileEx(path, STGM_WRITE | STGM_READ | STGM_SHARE_DENY_WRITE | STGM_CREATE, FILE_ATTRIBUTE_NORMAL, TRUE, nullptr, wil::out_param(outstream)));
+    return outstream;
+}
+
+winrt::hresult deploy_package_from_resource(ResourcePackageInfo resource, bool quiet) noexcept
+{
+    // Get package properties by loading the resource as a stream and reading the manifest.
+    wil::com_ptr<IStream> pkgstream;
+    std::unique_ptr<PackageProperties> properties;
+    try
+    {
+        pkgstream = get_resource_stream(resource.Id, resource.ResourceType);
+        properties = get_package_properites_from_stream(pkgstream);
+    }
+    catch (wil::ResultException ex)
+    {
+        return ex.GetErrorCode();
+    }
+
+    // Skip non-applicable architectures.
+    if (!is_architecture_applicable(properties->architecture))
+    {
+        return S_OK;
+    }
+
+
+    wchar_t package_filename[MAX_PATH];
+    RETURN_LAST_ERROR_IF(0 == GetTempFileName(std::filesystem::temp_directory_path().c_str(), L"PRP", 0u, package_filename));
+
+    if (!quiet)
+    {
+        printf("Package Full Name: %ls\n", properties->full_name.get());
+        printf("Temp package path: %ls\n", package_filename);
+    }
+
+    // Write the package to a temp file. The PackageManager APIs require a Uri.
+    wil::com_ptr<IStream> outstream;
+    try
+    {
+        outstream = open_temp_file(package_filename);
+    }
+    catch (wil::ResultException ex)
+    {
+        return ex.GetErrorCode();
+    }
+ 
+    ULARGE_INTEGER streamSize = { 0 };
+    RETURN_IF_FAILED(::IStream_Size(pkgstream.get(), &streamSize));
+    RETURN_IF_FAILED(pkgstream.get()->CopyTo(outstream.get(), streamSize, nullptr, nullptr));
+    RETURN_IF_FAILED(outstream->Commit(STGC_OVERWRITE));
+    outstream.reset();
+
+    // Add the package
+    Uri packageUri{ package_filename };
+    hresult add_result = add_package(packageUri);
+    if (!quiet)
+    {
+        printf("Package deployment result: 0x%08x\n", add_result.value);
+    }
+    RETURN_IF_FAILED((HRESULT)add_result);
+
+    // Best-effort provision
+    // Provisioning is expected to fail if the program is not run elevated or the user is not admin.
+    hresult provision_result = provision_package(properties->family_name.get());
+    if (!quiet)
+    {
+        printf("Provisioning result: 0x%08x\n", provision_result.value);
+    }
+    LOG_IF_FAILED((HRESULT)provision_result);
+
+    // Cleanup the temp file
+    LOG_IF_WIN32_BOOL_FALSE(::DeleteFile(package_filename));
```

Comment:
Add a `wil::scope_exit` to do this, or use @riverar 's temp-file-delete-on-close model.

Reply:
> Overall a good start. Are there tests or simulations to verify behavior in the CI/CD pipeline?

Tests are coming in an update a little later this week. Testing has been tricky because the installer is intended to be built with real packages and we don't want to be messing with the configuration of reunion on the dev machine when doing inner loop testing of the installer. So to work around that and the plan is to build 3 test packages (one for each framework architecture) and the default project will target including those test packages when no override header is present. The tests for the will the run the installer and verify the test packages are installed and then remove the test packages. So if you download the project and build the installer, you'll get an installer that installs just-built test packages that don't do anything to the system. That will allow inner loop testing of the installer and for tests to validate that the installer functions as expected. 

Then to produce real packages or packages for a targeted build, the override header will be included to target different packages. 

So pipeline is expected to be:
1) build the installer (using test packages)
2) test the installer
3) generate override header
4) generate the build's real reunion packages
5) build the installer using override header to target the real packages
6) sign the installer

Reply:
Seems like this tool could be used to bundle any set of MSIXs in any order though, right?  That is, the installer could find a little file within itself with a list of package descriptions (not baked into code), run the mini-applicability on the data in that file, and call the package manager APIs as necessary? Then you should be able to test that in isolation with completely bogus test packages now, no additional signing required.

@EHO-makai or @BenJKuhn, do we have a git large file storage location available to us somewhere? Committing large binaries to the repo is definitely not good, but somewhere we need packages available so everybody who's running the test suite locally can get them.

Reply:
Fixed to use scope_exit. Currently the installer project looks for a specific header with different package paths for identifying which packages to include, otherwise it includes some test-signed dummy packages it builds. The installer tests assume the dummy packages and verify their installation and do cleanup of those packages.

It can be used to bundle any set of MSIXes in any order, though there is no dependency mappings since this is intended to explicitly install frameworks in addition to main packages. For dependencies the order would potentially matter, but shouldn't be an issue if frameworks are registered first. If we want to move to only installing the main package and then adding other dependency packages it would change the install logic a bit. Dependency support could be added, though I don't think it is required at this time. I could add a script step that reads a data file and creates the header. That's essentially what the build pipeline needs to do - whether it's creating a data file or a header it still has to indicate the location of the packages to install. 

I'd want to carefully think about how to structure that process if it is to do some code generation during build, but I think what it has now is sufficient and I can update it as we discover pipeline issues or needs.

It doesn't need any packages in order to test the installer functionality. It will create dummy packages to verify deployment and the installer works as expected.